build failed on ubuntu 18.04 LTS(docker image)

Hi. I'm using a python library called wcmatch, and I'm having a build problem with the related dependency item, backrefs.

I think the problem appears in ubuntu 18.04.
And The error message looks like this:

What did I do wrong?
```
/# pip3 install backrefs
...<skipped>...
Skipping: found /tmp/pip-build-51bzznq8/backrefs/tools/unicodedata/9.0.0/BidiBrackets.txt
 ========Unicode Tables========
    Building: General Category
    Building: Blocks
    Traceback (most recent call last):
      File "/tmp/pip-build-51bzznq8/backrefs/setup.py", line 69, in generate_unicode_table
        'backrefs', 'uniprops', 'unidata'
      File "/tmp/pip-build-51bzznq8/backrefs/tools/unipropgen.py", line 1233, in build_tables
        build_unicode_property_table(output)
      File "/tmp/pip-build-51bzznq8/backrefs/tools/unipropgen.py", line 1219, in build_unicode_property_table
        gen_properties(output)
      File "/tmp/pip-build-51bzznq8/backrefs/tools/unipropgen.py", line 1045, in gen_properties
        gen_blocks(files['blk'], ascii_props, append, prefix)
      File "/tmp/pip-build-51bzznq8/backrefs/tools/unipropgen.py", line 173, in gen_blocks
        for line in uf:
      File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
        return codecs.ascii_decode(input, self.errors)[0]
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 59: ordinal not in range(128)

    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-51bzznq8/backrefs/setup.py", line 91, in <module>
        generate_unicode_table()
      File "/tmp/pip-build-51bzznq8/backrefs/setup.py", line 78, in generate_unicode_table
        assert not fail, "Failed uniprops.py generation!"
    AssertionError: Failed uniprops.py generation!
```

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,4 +1,5 @@
 build: false
+
 environment:
   matrix:
     - PYTHON: "C:/Python34"
@@ -11,22 +12,25 @@ environment:
       TOXENV: "py37"
     - PYTHON: "C:/Python36"
       TOXENV: "lint"
+
 init:
   - "ECHO %TOXENV%"
   - "ECHO %PYTHON%"
   - ps: "ls C:/Python*"
+
 install:
-  - "%PYTHON%/Scripts/pip.exe install -U setuptools"
-  - "%PYTHON%/Scripts/pip.exe install virtualenv"
-  - "%PYTHON%/Scripts/pip.exe install tox"
-  - "%PYTHON%/Scripts/pip.exe install codecov"
   - "SET PATH=%PYTHON%;%PYTHON%\\Scripts;%PATH%"
+  - "python -m pip install -U pip setuptools wheel"
+  - "python -m pip install -U virtualenv tox codecov"
+
 test_script:
   - pip --version
   - virtualenv --version
   - tox --version
   - tox
+
 after_test:
-  - if not ("%TOXENV%" == "lint") codecov
+  - if not ("%TOXENV%" == "lint") if not ("%TOXENV%" == "documents") codecov
+
 matrix:
   fast_finish: true

--- a/backrefs/__meta__.py
+++ b/backrefs/__meta__.py
@@ -186,5 +186,5 @@ def parse_version(ver, pre=False):
     return Version(major, minor, micro, release, pre, post, dev)
 
 
-__version_info__ = Version(4, 0, 0, "final")
+__version_info__ = Version(4, 0, 1, "final")
 __version__ = __version_info__._get_canonical()

--- a/backrefs/bregex.py
+++ b/backrefs/bregex.py
@@ -119,15 +119,15 @@ def _is_replace(obj):
 
 
 def _apply_replace_backrefs(m, repl=None, flags=0):
-        """Expand with either the `ReplaceTemplate` or compile on the fly, or return None."""
+    """Expand with either the `ReplaceTemplate` or compile on the fly, or return None."""
 
-        if m is None:
-            raise ValueError("Match is None!")
-        else:
-            if isinstance(repl, ReplaceTemplate):
-                return repl.expand(m)
-            elif isinstance(repl, (str, bytes)):
-                return _bregex_parse._ReplaceParser().parse(m.re, repl, bool(flags & FORMAT)).expand(m)
+    if m is None:
+        raise ValueError("Match is None!")
+    else:
+        if isinstance(repl, ReplaceTemplate):
+            return repl.expand(m)
+        elif isinstance(repl, (str, bytes)):
+            return _bregex_parse._ReplaceParser().parse(m.re, repl, bool(flags & FORMAT)).expand(m)
 
 
 def _apply_search_backrefs(pattern, flags=0):

--- a/docs/src/markdown/changelog.md
+++ b/docs/src/markdown/changelog.md
@@ -1,5 +1,9 @@
 # Changelog
 
+## 4.0.1
+
+- **FIX**: Ensure that when generating the Unicode property tables, that the property files are read in with `UTF-8` encoding.
+
 ## 4.0.0
 
 - **NEW**: Drop support for new features in Python 2. Python 2 support is limited to the 3.X.X series and will only receive bug fixes up to 2020. All new features moving forward will be on the 4.X.X series and will be for Python 3+ only.

--- a/requirements/test.txt
+++ b/requirements/test.txt
@@ -1,4 +1,4 @@
 pytest
 pytest-cov
 mock
-regex
+regex<=2018.11.22

--- a/tools/unipropgen.py
+++ b/tools/unipropgen.py
@@ -169,7 +169,7 @@ def gen_blocks(output, ascii_props=False, append=False, prefix=""):
         max_range = MAXASCII if ascii_props else MAXUNICODE
         formatter = bytesformat if ascii_props else uniformat
 
-        with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'Blocks.txt'), 'r') as uf:
+        with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'Blocks.txt'), 'r', 'utf-8') as uf:
             for line in uf:
                 if not line.startswith('#'):
                     data = line.split(';')
@@ -221,7 +221,7 @@ def gen_ccc(output, ascii_props=False, append=False, prefix=""):
     """Generate `canonical combining class` property."""
 
     obj = {}
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'DerivedCombiningClass.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'DerivedCombiningClass.txt'), 'r', 'utf-8') as uf:
         for line in uf:
             if not line.startswith('#'):
                 data = line.split('#')[0].split(';')
@@ -275,13 +275,13 @@ def gen_scripts(
     obj = {}
     obj2 = {}
     aliases = {}
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'PropertyValueAliases.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'PropertyValueAliases.txt'), 'r', 'utf-8') as uf:
         for line in uf:
             if line.startswith('sc ;'):
                 values = line.split(';')
                 aliases[format_name(values[1].strip())] = format_name(values[2].strip())
 
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, file_name_ext), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, file_name_ext), 'r', 'utf-8') as uf:
         for line in uf:
             if not line.startswith('#'):
                 data = line.split('#')[0].split(';')
@@ -297,7 +297,7 @@ def gen_scripts(
 
                     obj2[ext].extend(span)
 
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, file_name), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, file_name), 'r', 'utf-8') as uf:
         for line in uf:
             if not line.startswith('#'):
                 data = line.split('#')[0].split(';')
@@ -367,7 +367,7 @@ def gen_enum(file_name, obj_name, output, field=1, notexplicit=None, ascii_props
     """Generate generic enum."""
 
     obj = {}
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, file_name), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, file_name), 'r', 'utf-8') as uf:
         for line in uf:
             if not line.startswith('#'):
                 data = line.split('#')[0].split(';')
@@ -414,7 +414,7 @@ def gen_age(output, ascii_props=False, append=False, prefix=""):
 
     obj = {}
     all_chars = ALL_ASCII if ascii_props else ALL_CHARS
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'DerivedAge.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'DerivedAge.txt'), 'r', 'utf-8') as uf:
         for line in uf:
             if not line.startswith('#'):
                 data = line.split('#')[0].split(';')
@@ -465,7 +465,8 @@ def gen_nf_quick_check(output, ascii_props=False, append=False, prefix=""):
     categories = []
     nf = {}
     all_chars = ALL_ASCII if ascii_props else ALL_CHARS
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'DerivedNormalizationProps.txt'), 'r') as uf:
+    file_name = os.path.join(HOME, 'unicodedata', UNIVERSION, 'DerivedNormalizationProps.txt')
+    with codecs.open(file_name, 'r', 'utf-8') as uf:
         for line in uf:
             if not line.startswith('#'):
                 data = line.split('#')[0].split(';')
@@ -530,7 +531,7 @@ def gen_binary(table, output, ascii_props=False, append=False, prefix=""):
     )
     binary = {}
     for filename, include in binary_props:
-        with open(os.path.join(HOME, 'unicodedata', UNIVERSION, filename), 'r') as uf:
+        with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, filename), 'r', 'utf-8') as uf:
             for line in uf:
                 if not line.startswith('#'):
                     data = line.split('#')[0].split(';')
@@ -548,7 +549,7 @@ def gen_binary(table, output, ascii_props=False, append=False, prefix=""):
                         continue
                     binary[name].extend(span)
 
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'CompositionExclusions.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'CompositionExclusions.txt'), 'r', 'utf-8') as uf:
         name = 'compositionexclusion'
         for line in uf:
             if not line.startswith('#'):
@@ -565,7 +566,7 @@ def gen_binary(table, output, ascii_props=False, append=False, prefix=""):
                 binary[name].extend(span)
                 binary['full' + name].extend(span)
 
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'UnicodeData.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'UnicodeData.txt'), 'r', 'utf-8') as uf:
         name = 'bidimirrored'
         for line in uf:
             data = line.strip().split(';')
@@ -613,7 +614,7 @@ def gen_bidi(output, ascii_props=False, append=False, prefix=""):
 
     bidi_class = {}
     max_range = MAXASCII if ascii_props else MAXUNICODE
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'UnicodeData.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'UnicodeData.txt'), 'r', 'utf-8') as uf:
         for line in uf:
             data = line.strip().split(';')
             if data:
@@ -828,7 +829,7 @@ def gen_alias(enum, binary, output, ascii_props=False, append=False, prefix=""):
         'catalog', 'enumerated', 'numeric', 'miscellaneous'
     )
 
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'PropertyAliases.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'PropertyAliases.txt'), 'r', 'utf-8') as uf:
         div = False
         capture = False
         name = None
@@ -867,7 +868,7 @@ def gen_alias(enum, binary, output, ascii_props=False, append=False, prefix=""):
                     for d in data[1:]:
                         alias[name][d] = data[0]
 
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'PropertyValueAliases.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'PropertyValueAliases.txt'), 'r', 'utf-8') as uf:
         for line in uf:
             m = alias_re.match(line)
             if m:
@@ -1014,7 +1015,7 @@ def gen_properties(output, ascii_props=False, append=False):
     all_chars = ALL_ASCII if ascii_props else ALL_CHARS
     table = {'l': {'c': []}}
     itable = {'l': {}}
-    with open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'UnicodeData.txt'), 'r') as uf:
+    with codecs.open(os.path.join(HOME, 'unicodedata', UNIVERSION, 'UnicodeData.txt'), 'r', 'utf-8') as uf:
         for line in uf:
             data = line.strip().split(';')
             if data:

--- a/tox.ini
+++ b/tox.ini
@@ -5,7 +5,7 @@ envlist =
     spelling
 
 [testenv]
-passenv = LANG
+passenv = *
 deps=
     -rrequirements/project.txt
     -rrequirements/test.txt
@@ -14,7 +14,7 @@ commands=
     {envbindir}/coverage html -d {envtmpdir}/coverage
 
 [testenv:documents]
-basepython = python3.6
+passenv = *
 deps=
     -rrequirements/docs.txt
 commands=
@@ -22,6 +22,7 @@ commands=
     {envbindir}/pyspelling
 
 [testenv:lint]
+passenv = *
 deps=
     -rrequirements/project.txt
     -rrequirements/flake8.txt

